Allow the customization of the runner

gfontenot opened the pull request:
This makes it dead simple to customize xcodebuild.vim so that it works
with other libs such as vim-tmux-runner and vim-dispatch

https://github.com/christoomey/vim-tmux-runner
https://github.com/tpope/vim-dispatch

ping @christoomey because I'm referencing his fantastic library and want to
make sure I'm representing/using it properly.

Fixes #18

gfontenot committed:
Allow the customization of the runner

This makes it dead simple to customize xcodebuild.vim so that it works
with other libs such as vim-tmux-runner and vim-dispatch

https://github.com/christoomey/vim-tmux-runner
https://github.com/tpope/vim-dispatch

christoomey commented on doc/xcodebuild.txt:
```diff
@@ -18,6 +18,9 @@ xcodebuild. By default, xcodebuild.vim passes the `--color` flag to xcpretty.
 CONFIGURATION                             *xcodebuild-configuration*
 
+* Set the runner to use for the xcodebuild command (uses ! by default)
+ let g:xcodebuild_runner = 'VtrSendCommandToRunner'
```
:+1: The only other option is there is a bang form of the command, `VtrSendCommandToRunner!` that will open a pane if one is not currently attached. That's my preferred usage for these sort of interactions as it gets close to the "it just works" territory.

gfontenot replied:
I was literally going to ask you about this tomorrow. Fantastic note, I'll add that to this and the README.

christoomey commented on plugin/xcodebuild.vim:
```diff
@@ -80,6 +80,14 @@ function! s:sdk()
+  if exists('g:xcodebuild_runner')
+    return g:xcodebuild_runner
+  else
+    return '!'
```
The `!` feels like a magic ~~number~~ character here. Maybe extract and name this to something like `s:xcodebuild_default_background_runner`? You can pull the `let s:xcodebuild_default_background_runner = '!'` to the top of the script if you go that route.

gfontenot replied:
That's a good idea. I can extract the default flags used for xcpretty the same way. I'm going to do that in a different PR.

christoomey replied:
~~The `!` feels like a magic ~~number~~ character here. Maybe extract and name this to something like `s:xcodebuild_default_background_runner`? You can pull the `let s:xcodebuild_default_background_runner = '!'` to the top of the script if you go that route.~~

Commented more completely above with configuration suggestion.

gfontenot committed:
Use ! variant of VTR command

Using the ! version of `VtrSendCommandToRunner` will open a runner pane
if it doesn't already exist. That makes this a better default suggestion
for our documentation.

christoomey commented on plugin/xcodebuild.vim:
```diff
@@ -26,7 +26,7 @@ function! s:test()
 
 function! s:run_command(cmd)
-  execute '!' . a:cmd
+  execute s:runner() . ' ' . a:cmd
```
Big :+1: on adding support for configurable and async runners. This form will certainly work for the listed runners (shell, tmux-runner, and dispatch) as they all provide commands, but in some cases a user might want to run via a Vim function as the plugin they use only provides functions. Likewise, although you don't need it now, a template string approach can grow to support more options later if needed.

You can see the relevant examples in [vim-rspec](https://github.com/thoughtbot/vim-rspec/blob/2cf8c5002cb89eb8e9aabe59d25b89dc5afec88a/plugin/rspec.vim#L43-L47) and a more complex version in [vim-spec-runner](https://github.com/gabebw/vim-spec-runner/blob/2550392842e16a512fc42ef7ee761ad7180cf20e/plugin/spec-runner.vim#L165-L178).

This would likely mean having a command template string as the configuration point, rather than `g:xcodebuild_runner`. Something like:

``` vim
let  s:default_run_command = '!'

function! s:run_command(cmd)
  let run_cmd = substitute(s:runner_template(), "{cmd}", a:cmd, "g")
  execute run_cmd
endfunction

function! s:runner_template()
  if exists('g:xcodebuild_run_command')
    return g:xcodebuild_run_command
  else
    return s:default_run_command
  endif
endfunction
```

gfontenot replied:
am I correct in assuming that the default is missing a piece here? Should it be

``` vim
let s:default_run_command = '! {cmd}'
```

christoomey replied:
You are correct, sir!

gfontenot replied:
I dig this. Updated now.

christoomey commented:
Somehow I got [this notification](https://twitter.com/callouscommits/status/643236649225330688) before the github reference notification, which was first confusing, then excellent. Thanks for the kind words re: tmux-runner, and props on going full-on with your Vimscript adventures!

gfontenot commented:
Haha. That was from my dotfiles commit adding the plugin.

gfontenot committed:
Extract default values to script variables

Instead of hiding these away in the `else` statement for the accessors,
we can pull these up to the top and increase their visibility.

gfontenot committed:
Improve runner customization by using a template

This greatly improves the flexibility of the runner customization by
using a template string instead of a prefixed command. This allows for a
much larger range of customizations than the command technique would
have provided for.

Hat tip @christoomey
